hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -195,22 +233,90 @@ func (r *NodePoolReconciler) reconcile(ctx context.Context, hcluster *hyperv1.Ho
+		mcsServiceAccount.ImagePullSecrets = []corev1.LocalObjectReference{
+			{Name: controlplaneoperator.PullSecret(mcsServiceAccount.Namespace).Name},
+		}
+		return nil
+	})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.Log.Info("Reconciling MCS RoleBinding")
+	_, err = controllerutil.CreateOrUpdate(ctx, r.Client, mcsRoleBinding, func() error {
+		mcsRoleBinding.RoleRef = rbacv1.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "ClusterRole",
+			Name:     "edit",
+		}
+		mcsRoleBinding.Subjects = []rbacv1.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      mcsServiceAccount.Name,
+				Namespace: mcsServiceAccount.Namespace,
+			},
+		}
+		return nil
+	})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.Log.Info("Reconciling MCS Deployment")
+	_, err = controllerutil.CreateOrUpdate(ctx, r.Client, mcsDeployment, func() error {
+		return reconcileMCSDeployment(nodePool, mcsDeployment, mcsServiceAccount, releaseImage)
+	})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.Log.Info("Reconciling MCS Service")
+	_, err = controllerutil.CreateOrUpdate(ctx, r.Client, mcsService, func() error {
+		mcsService.Spec.Ports = []corev1.ServicePort{
+			{
+				Name:       "http",
+				Protocol:   corev1.ProtocolTCP,
+				Port:       80,
+				TargetPort: intstr.FromInt(8080),
+			},
+		}
+
+		mcsService.Spec.Selector = MachineConfigServerServiceSelector(nodePool.GetName())
+
+		mcsService.Spec.Type = corev1.ServiceTypeClusterIP
+		return nil
+	})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.Log.Info("Reconciling userdata Secret")
+	_, err = controllerutil.CreateOrUpdate(ctx, r.Client, userDataSecret, func() error {
+		disableTemplatingValue := []byte(base64.StdEncoding.EncodeToString([]byte("true")))
+		userDataValue := []byte(fmt.Sprintf(
+			`{"ignition":{"config":{"merge":[{"source":"http://%s/ign/%s","verification":{}}]},"security":{},"timeouts":{},"version":"3.1.0"},"networkd":{},"passwd":{},"storage":{},"systemd":{}}`,
+			hcluster.Status.IgnitionEndpoint,
+			nodePool.GetName()))
+		userDataSecret.Data = map[string][]byte{
+			"disableTemplating": disableTemplatingValue,
+			"value":             userDataValue,
 		}
-		mcs.Annotations[nodePoolAnnotation] = ctrlclient.ObjectKeyFromObject(nodePool).String()
-		mcs.Spec.ReleaseImage = nodePool.Spec.Release.Image
-		mcs.Spec.IgnitionService = nodePool.Spec.IgnitionService
 		return nil
-	}); err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed to reconcile machineConfigServer %q: %w",
-			ctrlclient.ObjectKeyFromObject(mcs).String(), err)
+	})
+	if err != nil {
+		return ctrl.Result{}, err
 	}
-	if mcs.Status.Version != targetVersion {
-		r.Log.Info("machineConfigServer status version does not match nodePool target version yet, waiting",
-			"machineConfigServer", ctrlclient.ObjectKeyFromObject(mcs).String())
+
+	if mcsDeployment.GetLabels()[payloadVersionAnnotation] == targetVersion || !DeploymentComplete(mcsDeployment) {
```

Comment:
when the versions match, there's nothing to do, but when progressing, should we `RequeueAfter`?

Reply:
We don't need to `RequeueAfter` since we are watching the the MCS deployment https://github.com/openshift/hypershift/pull/269/files#diff-c666a2a9dc48b0d3de33cc8fa3a054f853671f0428f9d56b263750dfc6d26e00R66